### PR TITLE
Remove experimental from Vector Crypto extensions

### PR DESCRIPTION
**Pull request by 4vtomat**

### CODE REVIEW

**Comment by asb** [comment]
Thanks for the patch, some very quick feedback and I'd highlight the first bullet as the most important, as this is potentially a blocker for graduating these extensions from experimental.

* My big concern with this would be the intrinsics - could you please comment on the status of their standardisation?
* While doing this change, it would make sense to update the header of RISCVInstrInfoZvk.td. While it doesn't explicitly say it describes an experimental version of the extension, it references 1.0.0-rc1 of the spec while presumably there's now a 1.0.0-final?
* Please update llvm/docs/RISCVUsage.rst
* Please add a release note to llvm/docs/ReleaseNotes.rst

**Comment by 4vtomat** [comment]
> Thanks for the patch, some very quick feedback and I'd highlight the first bullet as the most important, as this is potentially a blocker for graduating these extensions from experimental.
> 
> * My big concern with this would be the intrinsics - could you please comment on the status of their standardisation?
> * While doing this change, it would make sense to update the header of RISCVInstrInfoZvk.td. While it doesn't explicitly say it describes an experimental version of the extension, it references 1.0.0-rc1 of the spec while presumably there's now a 1.0.0-final?
> * Please update llvm/docs/RISCVUsage.rst
> * Please add a release note to llvm/docs/ReleaseNotes.rst

Currently I guess we're not finalized yet in https://github.com/riscv-non-isa/rvv-intrinsic-doc/pull/234.

> 
> Edit: When committing after review, the commit should be titled something like "[RISCV] Remove experimental from vector crypto extensions".

Oh, thanks, I forgot it~

**Comment by 4vtomat** [comment]
Since it's possible that **ISA spec** and **intrinsics spec** are not synchronized, so the updates add an dummy extension called **zexperimental**, once `-menable-experimental-extensions` is specified, the feature `zexperimental` is automatically added.
If specifying `let RequiredFeatures = ["Zexperimental"] ` is added in `.td` file, it will check if `zexperimental` exists, if not, the corresponding intrinsics would not be added, hence causing an **intrinsics undeclared** error.

**Comment by asb** [comment]
> Since it's possible that **ISA spec** and **intrinsics spec** are not synchronized, so the updates add an dummy extension called **zexperimental**, once `-menable-experimental-extensions` is specified, the feature `zexperimental` is automatically added.
> If `let RequiredFeatures = ["Zexperimental"] ` is added in `.td` file, it will check if `zexperimental` exists, if not, the corresponding intrinsics would not be added, hence causing an **intrinsics undeclared** error.

Thanks for looking into this. It's perhaps surprising we've gone so long without having the infrastructure to make an extension non-experimental without making its not-yet-finalized intrinsics non-experimental. Introducing a new required feature seems like a reasonable way of doing this, but introducing a dummy extension seems unnecessary (unless I'm missing something?).

I could see a solution involving a Clang flag (e.g. `-menable-experimental-intrinsics`), or perhaps a solution where the user must use a certain `#define` (I know riscv_vector.h is implemented using a pragma, but presumably that could still query defines). There may be other options too.

Another piece of info that would be helpful to better understand is what the process is for finalising intrinsic additions (@eopxd might know?). If it's just the case that we have to wait another month before moving an extension from experimental to experimental then maybe extra infro to separate experimental intrinsics from instructions is less interesting. If it's more indeterminate than that, we probably need it.

I'll put this on the agenda for Thursday's RISC-V LLVM sync-up call in the hope we can resolve it there.

File: llvm/lib/Support/RISCVISAInfo.cpp [hunk 1]
```diff
@@ -106,6 +106,8 @@ static const RISCVSupportedExtension SupportedExtensions[] = {
 
     {"zdinx", RISCVExtensionVersion{1, 0}},
 
+    {"zexperimental", RISCVExtensionVersion{1, 0}},
+
```
Comment (kito-cheng) [inline llvm/lib/Support/RISCVISAInfo.cpp:110]: I guess we don't really need this dummy extension in RISCVISAInfo?

**Comment by 4vtomat** [comment]
> > Since it's possible that **ISA spec** and **intrinsics spec** are not synchronized, so the updates add an dummy extension called **zexperimental**, once `-menable-experimental-extensions` is specified, the feature `zexperimental` is automatically added.
> > If `let RequiredFeatures = ["Zexperimental"] ` is added in `.td` file, it will check if `zexperimental` exists, if not, the corresponding intrinsics would not be added, hence causing an **intrinsics undeclared** error.
> 
> Thanks for looking into this. It's perhaps surprising we've gone so long without having the infrastructure to make an extension non-experimental without making its not-yet-finalized intrinsics non-experimental. Introducing a new required feature seems like a reasonable way of doing this, but introducing a dummy extension seems unnecessary (unless I'm missing something?).
> 
> I could see a solution involving a Clang flag (e.g. `-menable-experimental-intrinsics`), or perhaps a solution where the user must use a certain `#define` (I know riscv_vector.h is implemented using a pragma, but presumably that could still query defines). There may be other options too.
> 
> Another piece of info that would be helpful to better understand is what the process is for finalising intrinsic additions (@eopXD might know?). If it's just the case that we have to wait another month before moving an extension from experimental to experimental then maybe extra infro to separate experimental intrinsics from instructions is less interesting. If it's more indeterminate than that, we probably need it.
> 
> I'll put this on the agenda for Thursday's RISC-V LLVM sync-up call in the hope we can resolve it there.

Yeah, we don't need a new extension actually, thanks for pointing out!

File: llvm/lib/Support/RISCVISAInfo.cpp (hunk 1 above)
Comment (4vtomat) [reply llvm/lib/Support/RISCVISAInfo.cpp]: No, we don't~

**Review by topperc — COMMENTED** [review]
Do we need to add the "experimental" feature to RISCVFeatures.td? If the feature string shows up in the function attributes, won't the backend print that it doesn't recognize the feature name when it parses the string?

**Comment by 4vtomat** [comment]
> Do we need to add the "experimental" feature to RISCVFeatures.td? If the feature string shows up in the function attributes, won't the backend print that it doesn't recognize the feature name when it parses the string?

Yes, we do, I forgot it lol~

**Comment by 4vtomat** [comment]
This update add "experimental" to `RISCVFeatures.td`, also add a member variable `HasExperimental` to record whether the target has this feature.

**Comment by ebiggers** [comment]
Hi!  What's the status of this pull request?  Are there any major issues remaining?  It would be very useful to have the LLVM assembler officially support the vector crypto extensions.  I've been reviewing the [RISC-V vector crypto accelerated crypto routines for the Linux kernel](https://lore.kernel.org/linux-crypto/20231025183644.8735-1-jerry.shih@sifive.com), and currently they're emitting the instructions as bare `.inst`s in order to not rely on the assembler, which is quite ugly.  The GNU assembler already released the support for this a few months ago, in v2.41.

FYI @JerryShih and @phoebesv

**Comment by ebiggers** [comment]
FYI, I tried squashing the commits of this pull request and rebasing onto `main`, but there are conflicts in several files

File: llvm/docs/RISCVUsage.rst [hunk 2]
```diff
@@ -141,6 +141,23 @@ on support follow.
      ``Zve64f``       Supported
      ``Zve64d``       Supported
      ``Zvfh``         Supported
+     ``Zvbb``         Supported
+     ``Zvbc``         Supported
+     ``Zvkb``         Supported
+     ``Zvkg``         Supported
+     ``Zvkn``         Supported
+     ``Zvkn``         Supported
```
Comment (ebiggers) [inline llvm/docs/RISCVUsage.rst:149]: Zvkn is in this list twice

**Comment by ebiggers** [comment]
At https://github.com/ebiggers/llvm-project/tree/remove_experimental_from_vector_crypto I've rebased this pull request, squashed the commits, resolved conflicts, fixed the Zvkn duplication, and run `git clang-format`.  Note, it was necessary to change the type of `enum RVVRequire` from `uint16_t` to `unsigned int` because the next flag is now `1 << 16`, and to remove `experimental-` from some new files in `llvm/test/Transforms/` where references to the crypto extensions had been added.  @4vtomat can you update your pull request accordingly?  Feel free to reuse what I have.  Thanks!

**Comment by 4vtomat** [comment]
moved to the new [one](https://github.com/llvm/llvm-project/pull/74213)

**Pull request closed without merging**